File: firmware/esp32/counter.cpp
// firmware/esp32/counter.cpp
// The counter only ever goes up. A decrease is a bug, not a feature: the backend treats a
// count lower than the last accepted one as a rollback and rejects the report, so silently
// letting a count drop here would break every downstream check. There is deliberately no
// decrement, reset, or set-to API.
#include "counter.h"

#include "esp_log.h"

namespace meter {

static const char* TAG = "counter";

CounterEntry* Counter::find(const std::string& device_id) {
  for (CounterEntry& e : entries_) {
    if (e.device_id == device_id) return &e;
  }
  return nullptr;
}

const CounterEntry* Counter::find(const std::string& device_id) const {
  for (const CounterEntry& e : entries_) {
    if (e.device_id == device_id) return &e;
  }
  return nullptr;
}

void Counter::restore(const std::string& device_id, uint64_t count) {
  CounterEntry* e = find(device_id);
  if (e == nullptr) {
    entries_.push_back(CounterEntry{device_id, count});
    return;
  }
  if (count < e->count) {
    // Stale or corrupt persisted value. Keep the higher one.
    ESP_LOGW(TAG, "ignoring restore of %llu below live %llu for %s",
             (unsigned long long)count, (unsigned long long)e->count, device_id.c_str());
    return;
  }
  e->count = count;
}

void Counter::restore_sequence(uint64_t seq) {
  if (seq > sequence_) sequence_ = seq;
}

uint64_t Counter::increment(const std::string& device_id, uint64_t n) {
  CounterEntry* e = find(device_id);
  if (e == nullptr) {
    entries_.push_back(CounterEntry{device_id, n});
    return n;
  }
  // Saturate rather than wrap. A wrap would present as a rollback.
  if (e->count > UINT64_MAX - n) {
    e->count = UINT64_MAX;
  } else {
    e->count += n;
  }
  return e->count;
}

uint64_t Counter::get(const std::string& device_id) const {
  const CounterEntry* e = find(device_id);
  return e == nullptr ? 0 : e->count;
}

uint64_t Counter::next_sequence() {
  ++sequence_;
  return sequence_;
}

}  // namespace meter

File: gateway/counter.py
"""gateway/counter.py

Per-device monotonic usage counter plus a report sequence number. Mirrors
firmware/esp32/counter.cpp.

The count only ever goes up. A decrease is a bug, not a feature: the backend treats a
count below the last accepted one as a rollback and rejects the report, so a silent drop
here would break verification downstream. There is no decrement, reset or set API, and
restore() refuses to lower a live value.
"""

from __future__ import annotations


class Counter:
    def __init__(self) -> None:
        self._counts: dict[str, int] = {}
        self._sequence: int = 0

    def restore(self, device_id: str, count: int) -> None:
        """Load a persisted value. A value below the live one is ignored, not applied."""
        if count < 0:
            raise ValueError("count cannot be negative")
        if count >= self._counts.get(device_id, 0):
            self._counts[device_id] = count

    def restore_sequence(self, sequence: int) -> None:
        if sequence > self._sequence:
            self._sequence = sequence

    def increment(self, device_id: str, n: int = 1) -> int:
        if n < 1:
            raise ValueError("increment must be positive")
        new = self._counts.get(device_id, 0) + n
        self._counts[device_id] = new
        return new

    def get(self, device_id: str) -> int:
        return self._counts.get(device_id, 0)

    def snapshot(self) -> dict[str, int]:
        return dict(self._counts)

    @property
    def sequence(self) -> int:
        return self._sequence

    def next_sequence(self) -> int:
        """Advance once per report batch. Never reused, so a replay is detectable."""
        self._sequence += 1
        return self._sequence
